presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -786,50 +892,23 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
         }
     }
 
-    private static PrivilegeBag buildPrivilegeBag(
-            String databaseName,
-            String tableName,
-            HivePrincipal hivePrincipal,
-            Set<PrivilegeGrantInfo> privilegeGrantInfos)
-    {
-        ImmutableList.Builder<HiveObjectPrivilege> privilegeBagBuilder = ImmutableList.builder();
-        for (PrivilegeGrantInfo privilegeGrantInfo : privilegeGrantInfos) {
-            privilegeBagBuilder.add(
-                    new HiveObjectPrivilege(
-                            new HiveObjectRef(TABLE, databaseName, tableName, null, null),
-                            hivePrincipal.getPrincipalName(),
-                            hivePrincipal.getPrincipalType(),
-                            privilegeGrantInfo));
-        }
-        return new PrivilegeBag(privilegeBagBuilder.build());
-    }
-
-    private static boolean containsAllPrivilege(Set<PrivilegeGrantInfo> requestedPrivileges)
-    {
-        return requestedPrivileges.stream()
-                .anyMatch(privilege -> privilege.getPrivilege().equalsIgnoreCase("all"));
-    }
-
-    private Set<HivePrivilegeInfo> getTablePrivileges(HivePrincipal hivePrincipal, String databaseName, String tableName)
-    {
-        if (hivePrincipal.getPrincipalType() == ROLE) {
-            return getRolePrivileges(hivePrincipal, new HiveObjectRef(TABLE, databaseName, tableName, null, null));
-        }
-        return getUserPrivileges(hivePrincipal, new HiveObjectRef(TABLE, databaseName, tableName, null, null));
-    }
-
-    private Set<HivePrivilegeInfo> getRolePrivileges(HivePrincipal hivePrincipal, HiveObjectRef hiveObjectRef)
+    @Override
+    public Set<HivePrivilegeInfo> listTablePrivileges(String databaseName, String tableName, PrestoPrincipal principal)
     {
-        checkArgument(hivePrincipal.getPrincipalType() == ROLE, "Expected ROLE PrincipalType but found USER");
-
         try {
             return retry()
                     .stopOnIllegalExceptions()
                     .run("getListPrivileges", stats.getListPrivileges().wrap(() -> {
                         try (HiveMetastoreClient client = clientProvider.createMetastoreClient()) {
+                            Table table = client.getTable(databaseName, tableName);
```

Comment:
Method is invoked with triple `databaseName, tableName, principal`, while table is retrieved for tuple `databaseName, tableName`. In case of many roles this method will be called once per each role (`principal`) but the same `databaseName, tableName` casing many retrieval for the same table. 
This wouldn't be a problem if table was cached, but here we are under the cache so table is actually retrieved many times. 
This may make planing time to be long if there are many roles.